src/core/thread/topology.cpp

```diff
@@ -199,17 +199,26 @@ void Neighbor::RemoveAllForwardTrackingSeriesInfo(void)
 
 const char *Neighbor::StateToString(State aState)
 {
-    static const char *kStateStrings[] = {
-        "Invalid",        // kStateInvalid
-        "Restored",       // kStateRestored
-        "ParentReq",      // kStateParentRequest
-        "ParentRes",      // kStateParentResponse
-        "ChildIdReq",     // kStateChildIdRequest
-        "LinkReq",        // kStateLinkRequest
-        "ChildUpdateReq", // kStateChildUpdateRequest
-        "Valid",          // kStateValid
+    static const char *const kStateStrings[] = {
+        "Invalid",        // (0) kStateInvalid
+        "Restored",       // (1) kStateRestored
+        "ParentReq",      // (2) kStateParentRequest
+        "ParentRes",      // (3) kStateParentResponse
+        "ChildIdReq",     // (4) kStateChildIdRequest
+        "LinkReq",        // (5) kStateLinkRequest
+        "ChildUpdateReq", // (6) kStateChildUpdateRequest
+        "Valid",          // (7) kStateValid
     };
 
+    static_assert(0 == kStateInvalid, "kStateInvalid valie is incorrect");
+    static_assert(1 == kStateRestored, "kStateRestored valie is incorrect");
+    static_assert(2 == kStateParentRequest, "kStateParentRequest valie is incorrect");
+    static_assert(3 == kStateParentResponse, "kStateParentResponse valie is incorrect");
+    static_assert(4 == kStateChildIdRequest, "kStateChildIdRequest valie is incorrect");
+    static_assert(5 == kStateLinkRequest, "kStateLinkRequest valie is incorrect");
+    static_assert(6 == kStateChildUpdateRequest, "kStateChildUpdateRequest valie is incorrect");
+    static_assert(7 == kStateValid, "kStateValid valie is incorrect");
+
     return static_cast<uint8_t>(aState) < OT_ARRAY_LENGTH(kStateStrings) ? kStateStrings[aState] : "Unknown";
```

Comment:
We should be consistent on how we handle larger-than-expected values.

Reply:
Let me move `Stringify` to `common/string.hpp` and then we can use it everywhere.

I expect this may add a bit of code size overhead (doing the check against array length) but I think it may be worth the safety it adds....and it adds consistency. 😃

Reply:
I tried the model where we use `Stringify()` everywhere which practically does add the array length check everywhere in the core.

Its impact on code size is much more than what I expected:
```
  ot-cli-ftd_1.2               -32bdcae  461296  780   68872  530948
                               +038833e  461688  780   68872  531340
                               +/-       +392    0     0      +392
  ot-cli-mtd_1.2               -32bdcae  372976  780   50312  424068
                               +038833e  373368  780   50312  424460
                               +/-       +392    0     0      +392
  ot-ncp-ftd_1.2               -32bdcae  444684  780   68504  513968
                               +038833e  445108  780   68504  514392
                               +/-       +424    0     0      +424
  ot-ncp-mtd_1.2               -32bdcae  358372  780   49952  409104
                               +038833e  358748  780   49952  409480
                               +/-       +376    0     0      +376
  ot-rcp_1.2                   -32bdcae  60460   564   20172  81196
```

Here is the link to run: https://github.com/abtink/openthread/runs/4708791524?check_suite_focus=true

I am leaning towards not making this change and let us be selective about what to do in each situation within core.
- I think in some cases (where we may get the `aState` from a public API), we may want to do the check.
- In some other cases where the use is limited to same file/module we can skip the check.
- The other option is to use `OT_ASSERT()` in some cases (but I expect it may add code size overhead as well when asserts are enabled).

Thoughts? @jwhui

Reply:
Here is my suggestion for a guideline on how to use/handle `ToString()` methods in core modules

- If the `enum` and `ToString()` method are used within the same module (same file) only or their use is limited to OT core modules, then we don't add any checks (no assert or array length check). Examples of this are things like `Mle::MessageTypeToString` or `Neighbor::StateToString()` ... which are only used within core.

- If the method may be used from a OT public API then we either protect it with `OT_ASSERT()` check in the API or check for array length. I would say we leave the decision for which model to use based the method and its use.
   - For example, `Mle::RoleToString()` we use the array length check, since it also used in `otThreadDeviceRoleToString()` and maybe more importantly from `Settings` where we can read it from non-volatile memory (which may potentially have invalid/corrupted value).
   - For `Joiner::StateToString()` which is used from `otJoinerStateToString()` we use `OT_ASSERT()` in `otJoinerStateToString()` implementation to check the value. This should be fine since we expect the `otJoinerStateToString()` to use `otJoinerState` enum.

I updated the PR now to follow this. 
Thoughts/Suggestions?

Reply:
SGTM!